### PR TITLE
add response header custom tag

### PR DESCRIPTION
**Pull request by huanghuangzym**
fix for https://github.com/envoyproxy/envoy/issues/29036

Commit Message: add response header custom tag api define
Additional Description: we want to add custom response header tag
Risk Level: 
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/type/tracing/v3/custom_tag.proto [hunk 1]
```diff
@@ -96,7 +96,10 @@ message CustomTag {
     // A request header custom tag.
     Header request_header = 4;
 
+    // A response header custom tag.
+    Header response_header = 5;
+
     // A custom tag to obtain tag value from the metadata.
-    Metadata metadata = 5;
+    Metadata metadata = 6;
```
Comment (wbpcode) [inline api/envoy/type/tracing/v3/custom_tag.proto:103]: Please don't update API in this way, we should never change ID of a exist field. And if there are no related implementation, please use `[#not-implemented-hide:]` to hide this API.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for your contribution. And a first comment to the API change is added. :)

File: api/envoy/type/tracing/v3/custom_tag.proto (hunk 1 above)
Comment (huanghuangzym) [reply api/envoy/type/tracing/v3/custom_tag.proto]: thanks , fix for not change proto id.   and  i will implemented later .  so should i need to add not-implemented-hide tag?

Comment (wbpcode) [reply api/envoy/type/tracing/v3/custom_tag.proto]: if you will add the implementation in this PR, then you needn't to do that.

**Comment by wbpcode** [comment]
please run the `ci/do_ci.sh check_and_fix_proto_format` in the dev container to fix proto format first. Thanks.

**Comment by huanghuangzym** [comment]
> please run the `ci/do_ci.sh check_and_fix_proto_format` in the dev container to fix proto format first. Thanks.

done  for format check

--- a/api/envoy/type/tracing/v3/custom_tag.proto
+++ b/api/envoy/type/tracing/v3/custom_tag.proto
@@ -17,7 +17,7 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#protodoc-title: Custom Tag]
 
 // Describes custom tags for the active span.
-// [#next-free-field: 6]
+// [#next-free-field: 7]
 message CustomTag {
   option (udpa.annotations.versioning).previous_message_type = "envoy.type.tracing.v2.CustomTag";

**Comment by wbpcode** [comment]
Please merge main and push again, it may could fix the CI.

**Comment by wbpcode** [comment]
friendly ping @adisuissa for API review.

**Comment by wbpcode** [comment]
Hah, I just realized that you have add implementation in this PR...

**Comment by huanghuangzym** [comment]
> Hah, I just realized that you have add implementation in this PR...

yeah, just add the class define to avoid mobile build ci error

**Comment by wbpcode** [comment]
Thanks. But if implementation is necessary, then please complete it or community cannot accept a partial patch.

**Comment by wbpcode** [comment]
Seems this is still WIP, so I convert it to a draft and feel free to mark it as ready when you complete it or think it's ready for a formal review.

**Comment by huanghuangzym** [comment]
> Seems this is still WIP, so I convert it to a draft and feel free to mark it as ready when you complete it or think it's ready for a formal review.

ok ,i will try to complete this work

File: api/envoy/type/tracing/v3/custom_tag.proto [hunk 2]
```diff
@@ -98,5 +98,8 @@ message CustomTag {
 
     // A custom tag to obtain tag value from the metadata.
     Metadata metadata = 5;
+
+    // A response header custom tag.
```
Comment (adisuissa) [inline api/envoy/type/tracing/v3/custom_tag.proto:102]: nit: please move this directly after `request_header` (before `metadata`).

**Review by adisuissa — COMMENTED** [review]
Thanks!
API LGTM, left a small nit.

File: api/envoy/type/tracing/v3/custom_tag.proto (hunk 2 above)
Comment (huanghuangzym) [reply api/envoy/type/tracing/v3/custom_tag.proto:99]: ok， i have fix it

**Review by adisuissa — COMMENTED** [review]
/lgtm api

File: source/common/stream_info/stream_info_impl.h [hunk 3]
```diff
@@ -279,6 +279,12 @@ struct StreamInfoImpl : public StreamInfo {
 
   const Http::RequestHeaderMap* getRequestHeaders() const override { return request_headers_; }
 
+  void setResponseHeaders(const Http::ResponseHeaderMap& headers) override {
+    response_headers_ = &headers;
+  }
+
+  const Http::ResponseHeaderMap* getResponseHeaders() const override { return response_headers_; }
```
Comment (wbpcode) [inline source/common/stream_info/stream_info_impl.h:286]: I actually doubt that could we add this method. Long time ago, I did some discussion with other maintainers, we agreed on that the stream info should own the data it store and we finally will remove the `request_headers_` here.

Comment (wbpcode) [reply source/common/stream_info/stream_info_impl.h:286]: See https://github.com/envoyproxy/envoy/pull/21932#issuecomment-1202581791

And cc @adisuissa

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for your contribution. In fact, it is acceptable for me to add a new `ResponseHeaderMap*` to the `CustomTagContext` and it is better than to add the reponse pointer to stream info.

Although I hope all these tracing-related code are protocol independent and could reused by non-HTTP proxying, But I have to admit that HTTP is special and it's OK to treat it as super VIP.

File: source/common/stream_info/stream_info_impl.h (hunk 3 above)
Comment (huanghuangzym) [reply source/common/stream_info/stream_info_impl.h:286]: it seems that ,we can change CustomTagContext add response header? but can not add repsonse header in stream info ?

Comment (wbpcode) [reply source/common/stream_info/stream_info_impl.h:286]: I think so, according to the previous discussion or comments.

Comment (huanghuangzym) [reply source/common/stream_info/stream_info_impl.h:286]: yeah ,if previous discussion has fixed conclusion,i will try another way like struct CustomTagContext {
  const TraceContext* trace_context;
  const StreamInfo::StreamInfo& stream_info;
  const  http::ResponseHeaderMap
};   is that ok?

Comment (wbpcode) [reply source/common/stream_info/stream_info_impl.h:286]: Please add a detailed comment to tell this is only used for HTTP and is a special case.

Comment (huanghuangzym) [reply source/common/stream_info/stream_info_impl.h:286]: ok, i will try to add,can reopen this pr?

**Pull request closed without merging**